lib/Dialect/QUIR/Transforms/MergeCircuitMeasures.cpp

```diff
@@ -0,0 +1,434 @@
+  auto newName2 = duplicateCircuit(rewriter, nextCircuitOp, symbolMap,
+                                   newNameTemplate, "-m");
+
+  // merge measurements
+  std::vector<Type> typeVec;
+  std::vector<Value> valVec;
+  initTypeAndValVectors(typeVec, valVec, measureOp, nextMeasureOp);
+
+  remapArguments(rewriter, valVec, callCircuitOp, nextCallCircuitOp, circuitOp,
+                 nextCircuitOp, nextMeasureOp);
+
+  // find return and update
+  auto returnOp = dyn_cast<quir::ReturnOp>(&circuitOp.back().back());
+  assert(returnOp && "quir.circuit must end end a quir.return");
+
+  rewriter.setInsertionPoint(measureOp);
+  auto mergedOp = rewriter.create<MeasureOp>(
+      measureOp.getLoc(), TypeRange(typeVec), ValueRange(valVec));
+
+  rewriter.replaceOp(measureOp, ResultRange(mergedOp.getOuts().begin(),
+                                            mergedOp.getOuts().end()));
+
+  llvm::SmallVector<Type> outputTypes;
+  llvm::SmallVector<Value> outputValues;
+
+  buildOutputs(outputTypes, outputValues, returnOp, measureOp, mergedOp);
+
+  auto newCallOp =
+      createNewCallOp(rewriter, outputTypes, outputValues, returnOp,
+                      callCircuitOp, circuitOp, newName1);
+
+  dropNextMeasure(rewriter, outputTypes, outputValues, nextCircuitOp,
+                  nextMeasureOp);
+
+  // dice the output so we can specify which results to replace
+  auto iterSep = newCallOp.result_begin() + callCircuitOp.getNumResults();
+  rewriter.replaceOp(callCircuitOp,
+                     ResultRange(newCallOp.result_begin(), iterSep));
+  rewriter.replaceOp(nextCallCircuitOp,
+                     ResultRange(iterSep, newCallOp.result_end()));
+
+  // delete the nextCircuit if it is now empty (starts with a return)
+  auto firstReturnOp = dyn_cast<quir::ReturnOp>(&nextCircuitOp.front().front());
+  if (firstReturnOp) {
+    symbolMap.erase(nextCircuitOp.getSymName());
+    rewriter.eraseOp(nextCircuitOp);
+  }
+}
+
+// This pattern matches on two MeasureOps that are only interspersed by
+// classical non-control flow ops and merges them into one measure op
+struct CallCircuitAndCallCircuitTopologicalPattern
+    : public OpRewritePattern<CallCircuitOp> {
+  explicit CallCircuitAndCallCircuitTopologicalPattern(
+      MLIRContext *ctx, llvm::StringMap<Operation *> &symbolMap)
+      : OpRewritePattern<CallCircuitOp>(ctx) {
+    _symbolMap = &symbolMap;
+  }
+
+  llvm::StringMap<Operation *> *_symbolMap;
+
+  LogicalResult matchAndRewrite(CallCircuitOp callCircuitOp,
+                                PatternRewriter &rewriter) const override {
+
+    // check to see if CallCircuit is inside a function and ignore as
+    // functions do not label their qubit physical ids this causes later
+    // getOperatedQubits to fail
+    auto funcOp = dyn_cast<func::FuncOp>(callCircuitOp->getParentOp());
+    if (funcOp && (SymbolRefAttr::get(funcOp).getLeafReference() != "main"))
+      return failure();
```

Comment:
This is here due to the way that gate definitions are processed with circuits enabled. 

Right now if we have a gate like this in the qasm:
```
gate sx q0 {
 U(1.57079632679, 0.0, 3.14159265359) q0;
}
```

without circuits it is represented as:

```
  func.func @sx(%arg0: !quir.qubit<1>) {
    %angle = quir.constant #quir.angle<1.57079632679> : !quir.angle<64>
    %angle_0 = quir.constant #quir.angle<0.000000e+00> : !quir.angle<64>
    %angle_1 = quir.constant #quir.angle<3.1415926535900001> : !quir.angle<64>
    quir.builtin_U %arg0, %angle, %angle_0, %angle_1 : !quir.qubit<1>, !quir.angle<64>, !quir.angle<64>, !quir.angle<64>
    return
  }
```

with circuits it becomes:

```
  func.func @sx(%arg0: !quir.qubit<1>) {
    quir.call_circuit @circuit_0(%arg0) : (!quir.qubit<1>) -> ()
    return
  }
  quir.circuit @circuit_0(%arg0: !quir.qubit<1>) {
    %angle = quir.constant #quir.angle<1.57079632679> : !quir.angle<64>
    %angle_0 = quir.constant #quir.angle<0.000000e+00> : !quir.angle<64>
    %angle_1 = quir.constant #quir.angle<3.1415926535900001> : !quir.angle<64>
    quir.builtin_U %arg0, %angle, %angle_0, %angle_1 : !quir.qubit<1>, !quir.angle<64>, !quir.angle<64>, !quir.angle<64>
    quir.return
  }
```

the `func.func` operations which define the gate are not being decorated with the qubit ids so I can not assume that they are present. I am verifying that the `callCircuitOp` are children of the main functions so that I can make the assumption that I am able to look up the qubit ids. An alternative would be to fail if that is not the case.

Reply:
Without the main test the assert in `lookupQubitIdHandleError_`  is triggered. 

https://github.com/openqasm/qe-compiler/blob/6f7302cbaf9b67ba8ae17f68e8d7bddca34d5e8b/lib/Dialect/QUIR/IR/QUIROps.cpp#L63-L74

Reply:
Why are we declaring/embedding these in a `func.func` in the first place? It seems like the `quir.circuit` is sufficient?

Reply:
The gates had been declared in `func.func` for use with `quir.call_gate`. This pre-dates the introduction of `quir.circuits`. When `quir.circuits` was added I did not attempt to replace the `func.func` at the time since we wanted to support enabling and disabling circuits so I put a call_circuit inside the gate `func` rather than full replacing the func with the `quir.circuit`. I'll see how much of a change would be required to replace with `func` with `quir.circuit`.